File: C/Sorting_Algorithms_in_C/QuickSort.c
#include<stdio.h>

void printarray(int *A, int n){
    for(int i = 0; i < n; i++){
        printf("%d ", A[i]);
    }
    printf("\n");
}

int partition(int A[], int low, int high){
    int pivot = A[low];
    int temp;
    int i = low + 1;
    int j = high;

    do {
        // Move 'i' to the right until we find an element larger than the pivot
        while(i <= high && A[i] < pivot) {
            i++;
        }
        // Move 'j' to the left until we find an element smaller than the pivot
        while(j >= low && A[j] > pivot) {
            j--;
        }
        // Swap A[i] and A[j] if 'i' is less than 'j'
        if(i < j){
            temp = A[i];
            A[i] = A[j];
            A[j] = temp;
        }
    } while(i < j);

    // Swap pivot element with A[j]
    temp = A[low];
    A[low] = A[j];
    A[j] = temp;

    return j; // Return pivot index
}

void quicksort(int A[], int low, int high){
    if(low < high){
        int partitionIndex = partition(A, low, high);
        // Recursively sort elements before and after partition
        quicksort(A, low, partitionIndex - 1);
        quicksort(A, partitionIndex + 1, high);
    }
}

int main(){
    int A[] = {3, 6, 44, 67, 23};
    int n = sizeof(A) / sizeof(A[0]);

    printf("Original array: ");
    printarray(A, n);

    quicksort(A, 0, n - 1);

    printf("Sorted array: ");
    printarray(A, n);

    return 0;
}

File: JAVA/Sorting Methods/QuickSort.java
public class QuickSort {
    public static int partition(int[] A,int p, int r){
        int temp;
        int x = A[r];
        int i = p-1;
        for (int j=p;j<=r-1;j++){
            if(A[j]<=x){
                i = i+1;
                temp = A[i];
                A[i] = A[j];
                A[j] = temp;
            }
        }
        temp = A[i+1];
        A[i+1] = A[r];
        A[r] = temp;
        return i+1;
    }
    public static int[] quickSort(int[] A,int p, int r ){
        if (p<r){
            int q= partition(A, p, r);
            quickSort(A, p,q-1);
            quickSort(A, q+1, r);
        }
        return A;
    }
    public static void main(String[] args) {
        System.out.println("Hello and welcome! This is a simple Test for the quick sort!");
        int []A = new int[]{9, 4, 6, 3, 2};
        A = quickSort(A, 0, A.length - 1);
        for (int i = 0; i < A.length; i++) {
            System.out.println("i = " + A[i]);
        }

    }
}